Add Probe support for Azure Container Instances

neilpeterson opened the pull request:
Issue - https://github.com/terraform-providers/terraform-provider-azurerm/issues/2916

I've taken my best stab on style. If I've missed any standards etc. happy to update.

katbyte commented on azurerm/resource_arm_container_group_test.go:
```diff
@@ -729,7 +813,7 @@ resource "azurerm_container_group" "test" {
       }
     }
-  }
+	}
```
Could we revert this?
```suggestion
  }
```

katbyte commented on azurerm/resource_arm_container_group_test.go:
```diff
@@ -605,7 +645,29 @@ resource "azurerm_container_group" "test" {
       "secureFoo"  = "secureBar"
       "secureFoo1" = "secureBar1"
-    }
+		}
```
terraform configs should use spaces not tabs, as such could we revert this?
```suggestion
    }
```

katbyte commented on azurerm/resource_arm_container_group_test.go:
```diff
@@ -605,7 +645,29 @@ resource "azurerm_container_group" "test" {
+		}
+
+		readiness_probe {
+			exec = ["cat","/tmp/healthy"]
```
Could we align these assignments?

katbyte commented on azurerm/resource_arm_container_group_test.go:
```diff
@@ -605,7 +645,29 @@ resource "azurerm_container_group" "test" {
-    }
+		}
+
+		readiness_probe {
```
An change this to be spaces not tabs here.

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+
+	}
+
+	if FailureThreshold := input.FailureThreshold; FailureThreshold != nil {
```
Might i suggest using v instead of FailureThreshold here to better match the existing provider style?

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	}
+
+	if FailureThreshold := input.FailureThreshold; FailureThreshold != nil {
+		output["failure_threashold"] = input.FailureThreshold
```
This should be be `*FailureThreshold`, and with above suggestion:
```suggestion
		output["failure_threashold"] = *v
```

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	}
+
+	if InitialDelaySeconds := input.InitialDelaySeconds; InitialDelaySeconds != nil {
+		output["inital_delay_seconds"] = input.InitialDelaySeconds
```
Same here

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	}
+
+	if PeriodSeconds := input.PeriodSeconds; PeriodSeconds != nil {
+		output["period_seconds"] = input.PeriodSeconds
```
and here

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	}
+
+	if SuccessThreshold := input.SuccessThreshold; SuccessThreshold != nil {
+		output["sucess_threashold"] = input.SuccessThreshold
```
and here

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	}
+
+	if TimeoutSeconds := input.TimeoutSeconds; TimeoutSeconds != nil {
+		output["timeout_seconds"] = input.TimeoutSeconds
```
and here

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+
+	if HTTPGet := input.HTTPGet; HTTPGet != nil {
+		httpget := make(map[string]interface{})
+		httpget["path"] = *input.HTTPGet.Path
```
These values should be nil checked before being dereferenced

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+
+	if Exec := input.Exec; Exec != nil {
+		var exec []string
+		for _, v := range *input.Exec.Command {
```
We should nil check this

katbyte replied:
Could potentially use `utils.FlattenStringArray`

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1235,56 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+		httpget := make(map[string]interface{})
+		httpget["path"] = *input.HTTPGet.Path
+		httpget["port"] = *input.HTTPGet.Port
+		httpget["scheme"] = input.HTTPGet.Scheme
```
This should be converted to a string before assingment

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -762,6 +920,73 @@ func expandContainerVolumes(input interface{}) (*[]containerinstance.VolumeMount
+		if len(commands) > 0 {
+			execRaw := make([]string, 0, len(commands))
+
+			for _, command := range commands {
```
I think you could use `utils.ExpandStringArray` here

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -762,6 +920,73 @@ func expandContainerVolumes(input interface{}) (*[]containerinstance.VolumeMount
+	for _, p := range probeRaw {
+		probeConfig := p.(map[string]interface{})
+
+		if initialDelaySeconds := probeConfig["inital_delay_seconds"].(int); initialDelaySeconds > 0 {
```
_minor and not a blocker_ but you could use v here to simplify this a little:
```go
		if v := probeConfig["inital_delay_seconds"].(int); v > 0 {
			probe.InitialDelaySeconds = utils.Int32(int32(v))
		}
```

But totally optional change

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -661,6 +809,16 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
 
+		if v, ok := data["liveness_probe"]; ok {
+			livenessProbe := expandContainerProbe(v)
+			container.ContainerProperties.LivenessProbe = livenessProbe
```
_minor_ personally i would combine these two lines

```
container.ContainerProperties.LivenessProbe = expandContainerProbe(v)
```

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+							},
+						},
+
+						"readiness_probe": {
```
Given that these two properties share the exact same schema, would it make sense to create a helper function that returns the schema?

I am also not going to make the same comments above here.

neilpeterson replied:
@katbyte yep, I was wondering how to properly handle this. I would be happy to do this. Are you aware of any other places in the provider that have a similar function that I can use as guidance?

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
```
Could we validate this and set the portals default? (and document the default)

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
```
Could we validate this and set the portals default? (and document the default)

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
```
Could we validate this and set the portals default? (and document the default)

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
```
Could we validate this and set the portals default? (and document the default)

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
```
Could we validate this and set the portals default? (and document the default)

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+													Type:     schema.TypeInt,
+													Optional: true,
+													ForceNew: true,
+												},
```
Could we validate this with `validate.PortNumber`

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -307,6 +307,154 @@ func resourceArmContainerGroup() *schema.Resource {
+										Type:     schema.TypeList,
+										Optional: true,
+										ForceNew: true,
+										Elem:     &schema.Schema{Type: schema.TypeString},
```
Should we be validating this to prevent empty elements?

neilpeterson commented:
Thanks a bunch, @katbyte I will address each and also the linting issues picked up by Travis CI.

neilpeterson commented:
@katbyte I've addressed all feedback with the following notes:

- Because the two probe types share a scheme, use a helper function to populate properties ([link](https://github.com/terraform-providers/terraform-provider-azurerm/pull/3118/files#r268812783)). I am happy to do this assuming we can preserve the end user experience. Do you have a sample of where a similar helper function was used that I can draw from?
- Validate and add portal defailts. From what I can tell the Azure APIs are not enforcing default values nor a range of expected values ([link](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/containerinstance/resource-manager/Microsoft.ContainerInstance/stable/2018-10-01/containerInstance.json#L1108)). This fetaure is also not avaliable .via the Azure portal. I have added default taken from [this Kuberntes document](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes). I have contacted some MSFT folk to fund out what the expectation is here.

katbyte commented:
@neilpeterson,

Here is an example of a scheme function:
```go
func SchemaDevTestVirtualMachineInboundNatRule() *schema.Schema {
	return &schema.Schema{
		Type:     schema.TypeSet,
		Optional: true,
		// since these aren't returned from the API
		ForceNew: true,
		Elem: &schema.Resource{
			Schema: map[string]*schema.Schema{
				"protocol": {
					Type:     schema.TypeString,
					Required: true,
					ValidateFunc: validation.StringInSlice([]string{
						string(dtl.TCP),
						string(dtl.UDP),
					}, false),
				},

				"backend_port": {
					Type:         schema.TypeInt,
					Required:     true,
					ForceNew:     true,
					ValidateFunc: validate.PortNumber,
				},

				"frontend_port": {
					Type:     schema.TypeInt,
					Computed: true,
				},
			},
		},
	}
}
```

As for the validation if there are no easy win there don't bother yourself too much trying to find some, its not a blocker to merge without 🙂

neilpeterson commented:
@katbyte thanks a bunch. I've added the helper function and removed the defaults to be more in line with the API. When we sort the Azure end I will come back and add default values back in.

With that, I believe this is ready for another review. Thanks for your help.

katbyte commented on azurerm/helpers/azure/container_group.go:
```diff
@@ -0,0 +1,91 @@
+								ValidateFunc: validation.StringInSlice([]string{
+									string(containerinstance.HTTP),
+									string(containerinstance.HTTPS),
+								}, true),
```
Could we change this ti false? we are moving towards case sensitivity for properties like this

neilpeterson replied:
@katbyte I am struggling a bit with this one. The Azure API always returns `Http` or `Https`. When setting `ignoreCase` to `false` tests on this value begin to fail. 

My assumption is that these tests should work.

```
    liveness_probe {
      httpget {
        path   = "/"
        port   = 443
        scheme = "Http"
      }
      inital_delay_seconds = 1
      period_seconds       = 1
      failure_threashold   = 1
      sucess_threashold    = 1
      timeout_seconds      = 1
    }
```

and then:

```
resource.TestCheckResourceAttr(resourceName,"container.0.liveness_probe.0.httpget.0.scheme", "Http"),
```

However, it seems I am hitting [this](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/vendor/github.com/hashicorp/terraform/helper/validation/validation.go#L90), which returns:

```
testing.go:538: Step 0 error: config is invalid: azurerm_container_group.test: expected container.0.liveness_probe.0.httpget.0.scheme to be one of [http https], got Http
```

I am still investigating but if you have any advice, I would appreciate it :).

Thanks

neilpeterson replied:
Thanks, Tom. I figured something was going on here and had considered this approach. Appreciate the assist and validation.

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -661,6 +665,15 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
+
+		if v, ok := data["readiness_probe"]; ok {
+			readinessProbe := expandContainerProbe(v)
+			container.ContainerProperties.ReadinessProbe = readinessProbe
```
Could we make this one line for consistency with the above assignment?

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1084,61 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+		output["exec"] = *v.Command
+	}
+
+	if v := input.HTTPGet; v != nil {
```
_minor_ i would change this to
```suggestion
	if get := input.HTTPGet; v != nil {
```

so we can use it below

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1084,61 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+	if v := input.HTTPGet; v != nil {
+		httpget := make(map[string]interface{})
+
+		if v := input.HTTPGet.Path; v != nil {
```
```suggestion
		if v := get.Path; v != nil {
```

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1084,61 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+			httpget["path"] = *v
+		}
+
+		if v := input.HTTPGet.Port; v != nil {
```
```suggestion
		if v := get.Port; v != nil {
```

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1084,61 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+			httpget["port"] = *v
+		}
+
+		if input.HTTPGet.Scheme != "" {
```
```suggestion
		if get.Scheme != "" {
```

katbyte commented on azurerm/resource_arm_container_group.go:
```diff
@@ -1002,6 +1084,61 @@ func flattenContainerVolumes(volumeMounts *[]containerinstance.VolumeMount, cont
+		}
+
+		if input.HTTPGet.Scheme != "" {
+			httpget["scheme"] = input.HTTPGet.Scheme
```
```suggestion
			httpget["scheme"] = get.Scheme
```

tombuildsstuff commented on azurerm/helpers/azure/container_group.go:
```diff
@@ -0,0 +1,91 @@
+					},
+				},
+
+				"inital_delay_seconds": {
```
```suggestion
				"initial_delay_seconds": {
```

neilpeterson commented:
@tombuildsstuff, thanks a bunch. Changes all make sense, really appreciate the assistance in wrapping this up.

tombuildsstuff commented:
Tests pas:

<img width="471" alt="Screenshot 2019-04-08 at 05 00 35" src="https://user-images.githubusercontent.com/666005/55695971-41404c00-59bb-11e9-82a4-0edefae046ff.png">
